File: frontend/src/lib/aispm/defender-import.ts
'use client';

/**
 * Defender for Cloud Apps import API client (M3, issue #243).
 *
 * Mirrors the backend Pydantic schemas in `app/schemas/defender_import.py`
 * and the `/api/v1/discover/defender-import` router: a pasted export is
 * extracted + enriched server-side (see
 * `app/services/defender_import_service.py`'s `KNOWN_APPS` catalog), so
 * the client only needs to render what comes back. Delegates auth/refresh
 * to the shared ApiClient in `api.ts`, matching `lib/aispm/mcp-discovery.ts`.
 */

import { api } from '../api';

// ──────────────────────────────────────────────────────────────────────
// Types
// ──────────────────────────────────────────────────────────────────────

export type DefenderImportStatus = 'pending' | 'accepted' | 'dismissed';

export interface DiscoveredDefenderApp {
  id: string;
  canonical_name: string;
  name: string;
  category: string;
  users_count: number;
  data_volume_mb: number;
  risk_score: number;
  certifications: string[];
  compliance_status: Record<string, string>;
  status: DefenderImportStatus;
  promoted_vendor_id: string | null;
  imported_at: string;
}

export interface DefenderImportList {
  apps: DiscoveredDefenderApp[];
}

export interface DefenderImportRequest {
  raw_text?: string | null;
}

export interface DefenderImportResult {
  matched_lines: number;
  created: number;
  updated: number;
  apps: DiscoveredDefenderApp[];
}

export interface DefenderAppStatusUpdate {
  status: DefenderImportStatus;
}

export interface DefenderAppAcceptRequest {
  department?: string;
}

export interface DefenderAppAcceptResult {
  app: DiscoveredDefenderApp;
  vendor_id: string;
}

// ──────────────────────────────────────────────────────────────────────
// Sample export (paste-box placeholder / demo convenience)
// ──────────────────────────────────────────────────────────────────────

export const SAMPLE_DEFENDER_EXPORT = `ChatGPT Enterprise — 312 active users — 1.2 TB transferred
Grammarly Business — 178 active users — 340 GB transferred
Notion AI — 96 active users — 78 GB transferred
GitHub Copilot — 54 active users — 12 GB transferred
Otter.ai — 23 active users — 9 GB transferred
`;

// ──────────────────────────────────────────────────────────────────────
// Client
// ──────────────────────────────────────────────────────────────────────

const BASE = '/discover/defender-import';

export const defenderImportApi = {
  list: (): Promise<DefenderImportList> => api.request<DefenderImportList>(BASE),

  import: (payload: DefenderImportRequest): Promise<DefenderImportResult> =>
    api.request<DefenderImportResult>(`${BASE}/import`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),

  accept: (id: string, payload: DefenderAppAcceptRequest = {}): Promise<DefenderAppAcceptResult> =>
    api.request<DefenderAppAcceptResult>(`${BASE}/${id}/accept`, {
      method: 'POST',
      body: JSON.stringify(payload),
    }),

  updateStatus: (id: string, body: DefenderAppStatusUpdate): Promise<DiscoveredDefenderApp> =>
    api.request<DiscoveredDefenderApp>(`${BASE}/${id}`, {
      method: 'PATCH',
      body: JSON.stringify(body),
    }),

  dismiss: (id: string): Promise<DiscoveredDefenderApp> =>
    defenderImportApi.updateStatus(id, { status: 'dismissed' }),
};

File: backend/app/schemas/defender_import.py
"""Pydantic schemas for the Defender for Cloud Apps import router (#243).

Wire format for `app.models.defender_import.DiscoveredDefenderApp`. Enum
values mirror the ORM enum (snake_case), matching `schemas/mcp_discovery.py`.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.defender_import import DefenderImportStatus


class DiscoveredDefenderAppResponse(BaseModel):
    """An imported app candidate as returned by the API."""

    id: str
    canonical_name: str
    name: str
    category: str
    users_count: int
    data_volume_mb: int
    risk_score: int
    certifications: list[str]
    compliance_status: dict[str, str]
    status: DefenderImportStatus
    promoted_vendor_id: str | None = None
    imported_at: datetime

    class Config:
        from_attributes = True


class DefenderImportListResponse(BaseModel):
    """Imported app candidates, ready to render."""

    apps: list[DiscoveredDefenderAppResponse] = Field(default_factory=list)


class DefenderImportRequest(BaseModel):
    """A pasted Defender for Cloud Apps export (no real OCR in v1).

    An empty/omitted `raw_text` runs the extractor against the bundled
    seed export, matching the AISPM prototype's demo-data convention.
    """

    raw_text: str | None = Field(default=None, max_length=20_000)


class DefenderImportResponse(BaseModel):
    """Result of an import: matched lines plus the current candidate set."""

    matched_lines: int
    created: int
    updated: int
    apps: list[DiscoveredDefenderAppResponse]


class DefenderAppStatusUpdate(BaseModel):
    """Dismiss (or reset) an imported app candidate."""

    status: DefenderImportStatus


class DefenderAppAcceptRequest(BaseModel):
    """Optional overrides when accepting a candidate into the inventory."""

    department: str = Field(default="Unknown", min_length=1, max_length=255)


class DefenderAppAcceptResponse(BaseModel):
    """Result of accepting a candidate: the row plus the new vendor id."""

    app: DiscoveredDefenderAppResponse
    vendor_id: str
